homeassistant/components/dnsip/sensor.py

```diff
@@ -48,27 +54,71 @@ async def async_setup_platform(
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the DNS IP sensor."""
-    hostname = config[CONF_HOSTNAME]
-    name = config.get(CONF_NAME)
-    ipv6 = config[CONF_IPV6]
+    _LOGGER.warning(
+        "Configuration of the DNS IP platform in YAML is deprecated and will be "
+        "removed in Home Assistant 2022.4; Your existing configuration "
+        "has been imported into the UI automatically and can be safely removed "
+        "from your configuration.yaml file"
+    )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the dnsip sensor entry."""
+
+    hostname = entry.data[CONF_HOSTNAME]
+    name = entry.data[CONF_NAME]
 
-    if not name:
-        name = DEFAULT_NAME if hostname == DEFAULT_HOSTNAME else hostname
-    resolver = config[CONF_RESOLVER_IPV6] if ipv6 else config[CONF_RESOLVER]
+    resolver_ipv4 = entry.options.get(CONF_RESOLVER, entry.data[CONF_RESOLVER])
+    resolver_ipv6 = entry.options.get(
+        CONF_RESOLVER_IPV6, entry.data[CONF_RESOLVER_IPV6]
+    )
+    entities = []
+    if entry.data[CONF_IPV4]:
+        entities.append(
+            WanIpSensor(name, hostname, resolver_ipv4, False, entry, f"{hostname}_ipv4")
+        )
+    if entry.data[CONF_IPV6]:
+        entities.append(
+            WanIpSensor(name, hostname, resolver_ipv6, True, entry, f"{hostname}_ipv6")
+        )
 
-    async_add_devices([WanIpSensor(name, hostname, resolver, ipv6)], True)
+    async_add_entities(entities, update_before_add=True)
 
 
 class WanIpSensor(SensorEntity):
     """Implementation of a DNS IP sensor."""
 
-    def __init__(self, name: str, hostname: str, resolver: str, ipv6: bool) -> None:
+    _attr_icon = "mdi:web"
+
+    def __init__(
+        self,
+        name: str,
+        hostname: str,
+        resolver: str,
+        ipv6: bool,
+        entry: ConfigEntry,
+        unique_id: str,
+    ) -> None:
         """Initialize the DNS IP sensor."""
-        self._attr_name = name
+        self._attr_name = f"{name} IPv6" if ipv6 else name
+        self._attr_unique_id = f"{entry.entry_id}_{unique_id}"
```

Comment:
Currently we don't prevent the user from creating more than one of these for the hostname.  If they create two of them for the same hostname you will get a collision here.

Reply:
Actually since you are prefixing with the entry id, you won't.

Reply:
This is still a bit messy though.  I don't think there is a use case for allowing them to create multiple entries for the same hostname.   This integration is a rare case where the hostname the subject of the integration. Since the hostname stable for the life of the config entry it should be ok to use as the unique id

Reply:
I agree and it comes back to a previous discussion we had so I agree and therefore now set unique id on the config entry and rewrote this on the sensor.